### ISSUE TITLE
Redfish(LDAP):  Add support for LDAP configuration in bmcweb.

### ISSUE THREAD

**Issue by ratagupt** [openbmc/bmcweb#13]
SchemaFile:- https://redfish.dmtf.org/schemas/v1/AccountService.v1_3_1.json

In the above schema file there is a property named ActiveDirectory/LDAP

ActiveDirectory:- Would be used if ldap server is ActiveDirectory
LDAP:- would be used if ldap server is OpenLDAP/ other then AD.

LDAP /ActiveDirectory in redfish is of type of ExternalAccountProvider

Which Redfish Property(ActiveProperty/LDAP) to update it is same as D-Bus Property LDAPType.

Here is a mapping if LDAP Server is OpenLDAP/ActiveDirectory:
```
LDAP/ActiveDirectory:(RedfishProperty)

     AccountProviderType:- "LDAPService",
     ServiceEnabled :- "True,If LDAP needs to be used",
     ServiceAddress:- "Would be mapped to LDAPServerURI(D-Bus Property)",
     Authentication: 
            { "AutenticationType" : "UserNameandPassword"
              "UserName" : "LDAPBindDN (D-Bus Property)" 
               "Password" : "BindPassword "
            },
     LDAPService:
            {
                "SearchSettings" : 
                       {
                            "BaseDistinguishedNames" : "LDAPBaseDN(D-Bus Property)"
                            "UserNameAttribute" : "uid/cn"
                             "GroupsAttribute" : "memberof"
                        },
               },
```
Have not added the mapping for role-mapping, would add it through different story.

D-Bus interface which is used by the backend:- https://github.com/openbmc/phosphor-dbus-interfaces/blob/master/xyz/openbmc_project/User/Ldap/Config.interface.yaml

**Comment by edtanous** [comment]
Just so we're clear, we're not planning on supporting active directory as part of this work, right?  If you do want it supported, I suspect that is a separate task.

One part of this task that should also be considered is the LocalAccountAuth property on account service.  We should likely support that property at some point.

The above looks roughly correct for what I would expect.  I would recommend mocking up the GET requests with static data, and run it through the service validator first, then, once that's passing, start working on the implementation of pulling the data from the backend.  Feel free to develop however works for you, but mockup first, then real implementation tends to work well for me, and avoid rework.

**Comment by ratagupt** [comment]
Thanks @edtanous for your suggestion, I would try with mockup first then real.
I was planning to implement the ActiveDirectory with the same story.I am hoping only the property name is different in both the cases.
For non activeDirectory is it "LDAP" and for activeDirectory it would be "ActiveDirectory".

**Comment by ratagupt** [comment]
@edtanous Can we get the Redfish PATCH request format to remove the LDAP Configuration

I have couple of ideas for the same

1)  PATCH {"LDAP": null}   ===> Seems Redfish schema doesn't allow this.
<Property Name="LDAP" Type="AccountService.v1_3_0.ExternalAccountProvider" Nullable="false">

Redfish says that this property can't be null.

2) Can be nullify the sub properties of LDAP property under account service like below?

PATCH {"LDAP": {"AccountProviderType": null, "ServiceEnabled":null, "ServiceAddresses": null, "Authentication" : null, "LDAPService": null, "RemoteRoleMapping": null}}

Can I nullify the sub property of parent property which is marked as nullable = false? Hoping that we can do.

There is thread open in the redfishforum for the same
https://redfishforum.com/thread/172/delete-specific-value-list

Suppose the above two doesn't look good we have other option atleast we can disable the LDAP service by patching specific property under LDAP, the below option is just to disable the LDAP service, it is not removing the LDAP Configuration from the BMC.

PATCH {"LDAP": { "ServiceEnabled":false}}

So query from my side
=> What is the redfish URL to remove the LDAP configuration?

**Comment by edtanous** [comment]
What's wrong with empty strings instead of null?  Isn't that what we default to anyway?

**Comment by ratagupt** [comment]
1) If all the subproperties is null then we can delete the backend D-bus object and if the user does the GET request on the Account Service then the user would not get the LDAP property.

2) If we start patching all the properties with empty string then the user would get the LDAP property and its sub-properties having the empty string as part of GET request on the Account Service.

With the second option, we have to make some changes in the backend as we don't allow the null values for certain parameters like "Serviceaddress" and baseDN.
Backend verifies that the given address is valid or not and the baseDN can't be empty.

With the second option, the user has to do fill all the sub-properties with empty string.Currently, the code which is up there, it was done with the same thought process.

Curl query for 1st option.
====================
curl --insecure -k -H "X-Auth-Token: $bmc_token" -X PATCH https://${BMC_IP}/redfish/v1/AccountService -d '{"LDAP":{"AccountProvideType":"","ServiceEnabled":false,"ServiceAddresses": ["ldap://ipaddress"],"Authentication": {"AuthenticationType": "","Username": "","Password": ""},"LDAPService": {"SearchSettings": {"BaseDistinguishedNames": ["teststring"],"UsernameAttribute": "","GroupsAttribute": ""}}

Curl query for second option
========================
curl --insecure -k -H "X-Auth-Token: $bmc_token" -X PATCH https://${BMC_IP}/redfish/v1/AccountService -d '{"LDAP":{"AccountProvideType":"","ServiceEnabled":null,"ServiceAddresses": null,Authentication": null,"LDAPService": null}

What is your suggestion?

**Comment by edtanous** [comment]
I believe the 1st option is more compliant, but I'm far from an expert.  This is probably a better question for the redfish forum.

**Comment by ratagupt** [comment]
I have asked it in the redfish forum, seems I haven't get the reply but to way forward we can implement the same option1.
If in future when we get a reply we can make the change.

**Comment by ratagupt** [comment]
@edtanous : As we discussed on last community meeting, I had pushed the code which works only foir LDAP(Redfish property.)

I was planning to bring the configuration support for AD through Redfish.
**We have only one configuration on the backend.**

Having two properties(LDAP/AD) in redfish  requires that two configuration should be there on the backend.
                                         or
if we keep the same design where we have single config object in that case 
If user has configured LDAP and tries to configure the AD, then we an send a message saying that disable the LDAP service.
User would disable the service and when user configures the AD then its LDAP configuration will go away. 

I believe having multiple configuration support on backend support makes sense, even having multiple configuration on the backend we have to limit the user saying **that any moment of time only one can be active.**

Eg  User configures the LDAP and enables the service
      User configures the AD and enables the service.
**Now after the above configuration  only AD will be enabled and LDAP would be disabled.**

What is your views?

**Comment by edtanous** [comment]
@ratagupt As we've discussed, I would like to see LDAP merged before ActiveDirectory, so we're not trying to do too much work in parallel, and we can solve one problem at a time.  The above issue that you've mentioned is an issue that we should solve when we do the active directory work.  Functionally, this means that for Redfish patchsets, I will be reviewing the LDAP series first, then once that has been merged, I will review the Active Directory series.

I don't know enough about how the LDAP backend was implemented to know if one config that supports both, or two configs is better.  If two configs allows writing the parameters separately, that would be an advantage from a user standpoint, but I don't think it's required.  It should be noted, there are BMC implementations that exist that support both modes in parallel, with a selector to set the priority order between local accounts, LDAP accounts, and AD accounts.  Based on your questions, I suspect that you don't have those requirements, but it's good to know that it is possible.

From my perspective, the user facing interface needs to be consistent in its support, and provide the appropriate error codes if the user does something that is disallowed by the implementation, like enabling AD and LDAP at the same time.  How that's accomplished in the backend is a matter of implementation.

**Comment by ratagupt** [comment]
@edtanous :  
> As we've discussed, I would like to see LDAP merged before ActiveDirectory, so we're not trying to do too much work in parallel, and we can solve one problem at a time. The above issue that you've mentioned is an issue that we should solve when we do the active directory work. Functionally, this means that for Redfish patchsets, I will be reviewing the LDAP series first, then once that has been merged, I will review the Active Directory series.

I have uploaded the commits which is only for LDAP wrt to bmweb, It is not for AD. I thought, while you reviewing the LDAP series, I can start working on the AD.
As we have the AD support on the backend also but one type of confifuration at a time(either LDAP or AD).
If the commits are independent to each other(LDAP/AD) then we should not have problem, I was trying to get the expectation correct before start working on the AD.

>  It should be noted, there are BMC implementations that exist that support both modes in parallel, with a selector to set the priority order between local accounts, LDAP accounts, and AD accounts.

If we are talking about AMI based implementation then seems they have written their own LDAP client implementation which supports both(LDAP/AD) the configuration at the same time and we are leveraging the existing opensource implementation which limits us one configuration at a time.

> From my perspective, the user facing interface needs to be consistent in its support, and provide the appropriate error codes if the user does something that is disallowed by the implementation, like enabling AD and LDAP at the same time. How that's accomplished in the backend is a matter of implementation.
Ok, in that case  if user tries to PATCH the AD when LDAP is configured or vice versa,Give an error message saying that "Remove the Existing LDAP/AD configuration".
Then if user PATCHES with all empty for the LDAP/AD then we can delete the existing configuration and now user can PATCH other property(LDAP/AD).

**Comment by ratagupt** [comment]
Closing this as the support is there in the master.
